### PR TITLE
Major refactoring: subsonic service and controller

### LINKED ISSUE #207: Artist not shown
If I click on an artist (in the left artist pane), albums will show-up in a normal situation.
but if I Click an artist with only one album, no albums are showed.
(Same bug is in the Beta version)

### PR DESCRIPTION
**Pull request by Hyzual**
Hi !

**TL;DR** : All requests in `subsonic-service` have better error-reporting (e.g. server is down, wrong params, etc). We can also apply upgrades to users' localStorage-saved settings. Some dependencies versions have been changed, use `npm upgrade` and `bower upgrade` in the Jamstash dir.

This PR's aim is to refactor all the remaining requests in `subsonic-service.js` so they use `subsonicRequest()`. This enables us to send requests to Subsonic in a consistent manner : we always use the user-defined protocol, we always use angular's $http and most importantly we always have an error with http or subsonic errors ready to be displayed.

The idea was also to better separate responsibilities : the service focuses on querying Subsonic the right way and either returns an error or most often an array that contains the data we wanted (songs, albums, playlists, etc.). Then, the controller (`subsonic.js`) can set the scope up with the data, display an error (or not) and generally focus on all the display-related things. The result is lightweight functions in both the service and the controller.

Refactoring all this led me to modify some related services : I've added "array-mapping" functions to `model-service.js` that map all the songs in an array, or the podcasts, etc. I've also modified the way the DefaultSearchType setting works : we now store only an int. 0 for song, 1 for album, 2 for artist, since we already used those in `subsonic-service.js`.

Now because I changed that setting, I also added a way to automatically modify people's localStorage, because otherwise the `<select>` stayed blank until you went to the settings, changed the DefaultSearchType and saved the settings. `persistence-service` now manages loading and saving the settings from localStorage and will apply upgrades to them after loading. It's not perfect yet so the **first time** you load Jamstash v4.4.5, you will see a blank `<select>` to filter search results, but if you reload, it'll be filled. Having it work the first time requires a significant amount of work and further refactoring, that'll be for a future version.

On the topic of unit-tests, I am proud to announce that we're past 45 % of total code statements unit-tested ! I've added unit-tests to everything I touched, so we can be confident that changes we make don't break anything else (unless we wanted it to).

Other changes include adding a jscs rc file, see [jscs](http://jscs.info/). Basically it enables me to keep a consistent coding style. I've also added a `grunt-bump` config, so when you run `grunt bump` in the Jamstash dir, it will update all 3 `*.json` files to the next version. I've also upgraded `angular-locker` to v1.2.x, use `bower upgrade` in the Jamstash dir.

Some bugs were fixed with this refactoring. I believe #151 should be fixed, requires some more testing. Also, don't know if it was specific to my server but I could never display podcasts in Jamstash, this was fixed too (at least for my server :p). Also, when the `Volume` had never been set in localStorage, it caused a  jPlayer error "non-finite value" because it tried to assign undefined as a Volume. This should be fixed too, if it isn't defined the Volume will be initialized to 1.

On a final note, I've tried sending you emails to the address you use in commits but didn't get a response, did you get them ? Should I contact you that way or through some other channel ? I wanted to share with you what I was going to work on next.

Have a great weekend !

**[EDIT 2015-04-12]:** I've added to this PR 6 more commits to fix #207. It turns out, for every request that expects an array (of songs, artists, podcasts, playlists, etc), whenever there's only one element, Subsonic will return it as an array of 1 element when Madsonic will return it as an object. This caused our code to consider the response "empty" and to "fail silently" because we don't display a notification.
I've ensured we "convert" everything that should be an array into an array, so this should be fixed and unit-tested in v4.4.5

### CODE REVIEW

**Comment by tsquillario** [comment]
Shit man, I forgot about this! I'll get this merged in and add you to the repo!

**Comment by Hyzual** [comment]
Cool ! Thanks :)
